File: Microsoft Office Developer Documentation Team/Office-2010-Deploying-Multiple-Office-2010-3887/[C#]-Office 2010 Deploying Multiple Office 2010 Projects in One Package/C#/WordSmartTagAddIn/TemperatureSmartTag.cs
using System.Globalization;
using Word = Microsoft.Office.Tools.Word;
using System.Text.RegularExpressions;
using MSWord = Microsoft.Office.Interop.Word;
using Microsoft.Office.Tools.Word;

namespace WordSmartTagAddIn
{
  class TemperatureSmartTag : Word.ISmartTagExtension
  {
    const string TEMPERATURE_CELSIUS = "C";
    Word.Action temperatureAction = null;
    private Word.SmartTag smartTagDemo = null;

    public TemperatureSmartTag()
    {
      this.smartTagDemo = Globals.Factory.CreateSmartTag(
        "http://www.microsoft.com#TemperatureSmartTag", "Temperature", this);
      // Add a regular expression for the Temperature SmartTag recognition.
      smartTagDemo.Expressions.Add(new Regex(
        "\\s*(?'number'[-+]?[0-9]+)°?\\s?(?'temperature'(F|C|f|c))\\s*"));
      // Create an instance of the temperature action.
      temperatureAction = Globals.Factory.CreateAction("Convert");
      // Add the action to the Actions array for the smart tag.
      smartTagDemo.Actions = new Word.Action[] { temperatureAction };
      temperatureAction.BeforeCaptionShow += TemperatureAction_BeforeCaptionShow;
      temperatureAction.Click += TemperatureAction_Click;
    }

    public void TemperatureAction_BeforeCaptionShow(object sender, ActionEventArgs e)
    {
      string temperature = e.Properties.Read[
        "temperature"].ToUpper(CultureInfo.CurrentCulture);
      if (temperature == TEMPERATURE_CELSIUS)
      {
        ((Word.Action)sender).Caption = "Convert to Fahrenheit";
      }
      else
      {
        ((Word.Action)sender).Caption = "Convert to Celsius";
      }
    }

    public void TemperatureAction_Click(object sender, ActionEventArgs e)
    {
      double fahrenheit = 0;
      double celsius = 0;
      //Read the temperature type and value from the smart tag properties.
      string value = e.Properties.Read["number"];
      string temperature = e.Properties.Read[
        "temperature"].ToUpper(CultureInfo.CurrentCulture);
      string resultText = null;
      if (temperature == TEMPERATURE_CELSIUS)
      {
        celsius = System.Convert.ToDouble(
          value, NumberFormatInfo.CurrentInfo);
        fahrenheit = System.Math.Round((celsius * 9 / 5.0) + 32);
        resultText = string.Format(" {0}° F ", 
          fahrenheit.ToString("0", NumberFormatInfo.CurrentInfo));
      }
      else
      {
        fahrenheit = System.Convert.ToDouble(
          value, NumberFormatInfo.CurrentInfo);
        celsius = System.Math.Round((fahrenheit - 32) * 5 / 9.0);
        resultText = string.Format(" {0}° C ", 
          celsius.ToString("0", NumberFormatInfo.CurrentInfo));
      }
      e.Range.Collapse(MSWord.WdCollapseDirection.wdCollapseStart);
      e.Range.Delete(MSWord.WdUnits.wdCharacter, e.Text.Length);
      e.Range.InsertAfter(resultText);
    }

    public void Recognize(string text, Microsoft.Office.Interop.SmartTag.ISmartTagRecognizerSite site, Microsoft.Office.Interop.SmartTag.ISmartTagTokenList tokenList, SmartTagRecognizeContext context)
    {
    }

    public object ExtensionBase
    {
      get { return smartTagDemo; }
    }

    public Word.SmartTag Base
    {
      get { return smartTagDemo; }
    }
  }
}

File: Microsoft Office Developer Documentation Team/Office-2010-Deploying-Multiple-Office-2010-3887/[VB]-Office 2010 Deploying Multiple Office 2010 Projects in One Package/VB/WordSmartTagAddIn/TemperatureSmartTag.vb
Imports System.Globalization
Imports Word = Microsoft.Office.Tools.Word
Imports System.Text.RegularExpressions
Imports MSWord = Microsoft.Office.Interop.Word

Public Class TemperatureSmartTag
  Implements Word.ISmartTagExtension

  Const TEMPERATURE_CELSIUS As String = "C"
  Dim WithEvents temperatureAction As Word.Action
  Private smartTagDemo As Word.SmartTag

  Public Sub New()
    Me.smartTagDemo = Globals.Factory.CreateSmartTag(
      "http://www.microsoft.com#TemperatureSmartTag",
      "Temperature", Me)
    ' Add a regular expression for the Temperature SmartTag recognition.
    smartTagDemo.Expressions.Add(New Regex(
      "\s*(?'number'[-+]?[0-9]+)°?\s?(?'temperature'(F|C|f|c))\s*"))
    ' Create an instance of the temperature action.
    temperatureAction = Globals.Factory.CreateAction("Convert")
    ' Add the action to the Actions array for the smart tag.
    smartTagDemo.Actions = New Word.Action() {temperatureAction}
  End Sub

  Public Sub TemperatureAction_BeforeCaptionShow(ByVal sender As Object, ByVal e As ActionEventArgs) Handles temperatureAction.BeforeCaptionShow
    Dim temperature As String = e.Properties.Read(
      "temperature").ToUpper(CultureInfo.CurrentCulture)
    If temperature = TEMPERATURE_CELSIUS Then
      CType(sender, Word.Action).Caption = "Convert to Fahrenheit"
    Else
      CType(sender, Word.Action).Caption = "Convert to Celsius"
    End If
  End Sub

  Public Sub TemperatureAction_Click(ByVal sender As Object, ByVal e As ActionEventArgs) Handles temperatureAction.Click
    Dim fahrenheit As Double
    Dim celsius As Double
    'Read the temperature type and value from the smart tag properties.
    Dim value As String = e.Properties.Read("number")
    Dim temperature As String = e.Properties.Read(
      "temperature").ToUpper(CultureInfo.CurrentCulture)
    Dim resultText As String

    If temperature = TEMPERATURE_CELSIUS Then
      celsius = Convert.ToDouble(
        value, NumberFormatInfo.CurrentInfo)
      fahrenheit = Math.Round((celsius * 9 / 5) + 32)
      resultText = String.Format("{0}° F ",
        fahrenheit.ToString("0", NumberFormatInfo.CurrentInfo))
    Else
      fahrenheit = Convert.ToDouble(
        value, NumberFormatInfo.CurrentInfo)
      celsius = Math.Round((fahrenheit - 32) * 5 / 9)
      resultText = String.Format("{0}° C ",
        celsius.ToString("0", NumberFormatInfo.CurrentInfo))
    End If
    e.Range.Collapse(MSWord.WdCollapseDirection.wdCollapseStart)
    e.Range.Delete(MSWord.WdUnits.wdCharacter, e.Text.Length)
    e.Range.InsertAfter(resultText)
  End Sub

  Public ReadOnly Property Base As Word.SmartTag
    Get
      Return smartTagDemo
    End Get
  End Property

  Public ReadOnly Property ExtensionBase As Object Implements Microsoft.Office.Tools.IExtension.ExtensionBase
    Get
      Return smartTagDemo
    End Get
  End Property

  Public Sub Recognize(ByVal text As String, ByVal site As Microsoft.Office.Interop.SmartTag.ISmartTagRecognizerSite, ByVal tokenList As Microsoft.Office.Interop.SmartTag.ISmartTagTokenList, ByVal context As Microsoft.Office.Tools.Word.SmartTagRecognizeContext) Implements Microsoft.Office.Tools.Word.ISmartTagExtension.Recognize

  End Sub
End Class
